virtcontainers/qemu.go

```diff
@@ -1275,6 +1275,71 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
 	return nil
 }
 
+// Query QMP to find the PCI slot of a device, given its QOM path or ID
+func (q *qemu) qomGetSlot(qomPath string) (uint8, error) {
+	addr, err := q.qmpMonitorCh.qmp.ExecQomGet(q.qmpMonitorCh.ctx, qomPath, "addr")
+	if err != nil {
+		return 0xff, err
+	}
+	addrf, ok := addr.(float64)
+	// XXX going via float makes no real sense, but that's how
+	// JSON works, and we'll get away with it for the small values
+	// we have here
+	if !ok {
+		return 0xff, fmt.Errorf("addr QOM property of %q is %T not a number", qomPath, addr)
+	}
+	addri := uint8(addrf)
+
+	slotNum, funcNum := addri>>3, addri&0x7
+	if funcNum != 0 {
+		return 0xff, fmt.Errorf("Unexpected non-zero PCI function (%02x.%1x) on %q",
+			slotNum, funcNum, qomPath)
+	}
+
+	return slotNum, nil
+}
+
+// Query QMP to find a device's PCI path given its QOM path or ID
+// "PCI path" means the PCI slot number of each bridge leading to the
+// device, and finally the device itself, as a '/' separated list
```

Comment:
@dgibson Looks like this is the missing bit from `clh.go` ( and maybe cloud-hypervisor) that prevents your PR from working with cloud-hypervisor.  

Excuse my limited knowledge in the area. Can we derive the `PCI path` from the information already provided by cloud-hypervisor (predicted guest `BDF` for the hotplugged VFIO device and CLH only has one PCI bus)? Or you think we will need more information from cloud-hypervisor?

/cc @sboeuf Please feel free to share your thoughts. Thank you.

Reply:
Yes, this is the spot.  So, in general we cannot derive a PCI path from a BDF, but in this case we can.  The only reason clh is able to get away with using a bdf here is because it always places the device on the root bus, so domain and bus are always zero.  So, the PCI path is simply the slot number from the BDF.

If clh ever starts using bridges, we'll have to change this... but in that case clh will have to change its interface anyway, since it won't be able to reliably report a BDF in that case anyway.

Reply:
@dgibson Thanks for the explanation as always. Let's derive the `PCI path` from the BDF returned by clh for now. We will make it more generic when cloud-hypervisor changes its interface later. 

Can you please give it a try by updating the `clh.go` as you did for the debugging? Please also include a note on the assumptions we have here on cloud-hypervisor. Let's see whether that fixes the VFIO CI failure. Any other thoughts, please let me know.

Reply:
Yes, I plan to do that after rebasing on the PCI path generalization patches.